keps/draft-20180418-security-profile.md

```diff
@@ -0,0 +1,1516 @@
+and find security profile `saas-multitenancy-1.0-0` satisfies the needs.
+
+She creates a Kubernetes cluster with security profile `saas-multitenancy-1.0-0`
+Later on, she defines ResourceQuota as a template and
+use security profile's namespace population mechanism to
+populate the ResourceQuota object in users' namespaces.
+
+## Implementation
+
+### The Security Profile
+A _Security Profile_ is a set of Kubernetes manifest files, defining objects:
+
+- The Custom Resource of kind `SecurityProfile` containing rules for:
+  - Cluster bootstrapping configurations
+  - Namespace-scope policy object definitions
+- Cluster-scope object definitions
+- The `Deployment` objects to deploy controllers to enforce the rules
+  - Controllers watching namespaces and creates policy objects
+  - External admission control webhook
+  - Related configuration objects
+
+Please refer to [default-1.0-0](#the-default-profile) as an example.
+
+The Custom Resource `Security Profile` defines two sets of rules: bootstrapping rules and runtime rules
+
+#### Bootstrapping Rules
+
+The bootstrapping rules provide instructions for Kubernetes deployment tools
+to set up configurations correctly during cluster creating, upgrading/downgrading, including:
+
+- Command line flags for etcd, API server, controller manager, scheduler and kubelet
+- Locations and permissions of certificates, keys, configuration files
+- Security-relevant Operating System settings, for example
+  - Configurations for SELinux, AppArmor, seccomp
+  - Rules in iptables
+
+All of them are described as actionable rules which contains:
+
+- `name`: the unique name which identifies _a thing_ to be enforced;
+- `spec`: describes the desired state when _the thing_ identified by `name` is enforced.
+
+As security profiles are portable,
+the `name` must be globally unique across all security profiles,
+all clusters, and specific to define what to be enforced.
+The schema of `spec` is defined by `name`. When `name` is defined,
+the schema of `spec` is finalized, forever.
+Further change of schema will require a new unique name.
+
+With rules predefined,
+a security profile aware Kubernetes deployment tool should be able to
+download and inspect the selected security profile before creating/upgrading a cluster,
+and setup command line flags, config files properly according to the rules during cluster creation/upgrade.
+
+These rules can also be interpreted as a set of check conditions for tools to validate if the cluster has been setup properly.
+
+Please see [Predefined Rules](#predefined-rules) as a reference of the rules.
+
+As currently, there's no way (or limited) to inspect the host operating system from inside the cluster,
+these rules are only enforced by the deployers (or cluster lifecycle management system).
+
+#### Runtime Rules
+Runtime rules are enforced after the cluster is bootstrapped,
+by _enforcers_ (the controllers deployed with security profiles in the cluster).
```

Comment:
Would a runtime enforcer just alert on a rule violation? Or force move to a new state? or....?

Reply:
The runtime enforcers are currently enforcing by creating policy objects in newly created namespaces. There are multiple levels of enforcement:

1. Onetime enforcement when a namespace is created
2. Continuously monitoring all namespaces and detect violations
3. Continuously fixing detected violations

The proposal focuses on 1, and keep 2 and 3 out of scope. I will call this out.